### PR TITLE
MH-12821 Crop service

### PR DESCRIPTION
**Pull request by lwillmann**
The pull request is based on the closed pull request #178 . 

The workflow operation excutes ffmpeg to detect and remove black bars from videos.

I've taken @isabellakutger work and refactored some parts of her code and added a few Tests.

### CODE REVIEW

File: modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+    if (candidates.size() == 0) {
+      logger.info("No matching tracks available for cropping in workflow {}", workflowInstance);
+      return createResult(WorkflowOperationResult.Action.CONTINUE);
+    } else if (candidates.size() > 1) {
+      logger.info("Found more than one track to crop");
+    }
+
+    // start cropping all candidates in parallel
+    //List<Job> jobs = new ArrayList<>();
```
Comment (CGreweling) [inline modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java]: I think this unused codeline needs to be deleted? ;)

Comment (lwillmann) [reply modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java]: Definitely Christian ;-). 
Thanks a lot for the first review.

File: docs/guides/admin/docs/workflowoperationhandlers/cropvideo-woh.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Crop Workflow operation
+
+The plugin provides the workflow operation `crop-video`. This workflow operation excutes ffmpeg command `cropdetect`.
+`cropdetect` checks for black bars on the sides of the track of the workflow instance. If `cropdetect` is successfully,
```
Comment (doofy) [inline docs/guides/admin/docs/workflowoperationhandlers/cropvideo-woh.md]: successfully -> successful

Comment (lwillmann) [reply docs/guides/admin/docs/workflowoperationhandlers/cropvideo-woh.md]: Surely Michael ;-)
Thanks for the review.

**Comment by lkiesow** [comment]
@gregorydlogan, can you squash-merge this?
…and remembering the original pull request, you might want to test if this works on a multi-node set-up.

**Comment by gregorydlogan** [comment]
@lkiesow will try and remember.  And will double check on a multinode setup :)

**Comment by gregorydlogan** [comment]
For the record, this seems to be broken on the worker node:
```
org.apache.felix.resolver.reason.ReasonException: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=opencast-worker; type=karaf.feature; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; filter:="(&(osgi.identity=opencast-worker)(type=karaf.feature)(version>=8.0.0.SNAPSHOT)(version<=8.0.0.SNAPSHOT))" [caused by: Unable to resolve opencast-worker/8.0.0.SNAPSHOT: missing requirement [opencast-worker/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-services-processing-light-load; type=karaf.feature [caused by: Unable to resolve opencast-services-processing-light-load/8.0.0.SNAPSHOT: missing requirement [opencast-services-processing-light-load/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-videoeditor-workflowoperation; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-videoeditor-workflowoperation/8.0.0.SNAPSHOT: missing requirement [opencast-videoeditor-workflowoperation/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.workflow.handler.workflow)(version>=8.0.0)(!(version>=9.0.0)))"]]]
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
        at org.apache.felix.resolver.ResolverImpl.doResolve(ResolverImpl.java:392)
        at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:378)
        at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:332)
        at org.apache.karaf.features.internal.region.SubsystemResolver.resolve(SubsystemResolver.java:257)
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:388)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1025)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:964)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-worker/8.0.0.SNAPSHOT: missing requirement [opencast-worker/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-services-processing-light-load; type=karaf.feature [caused by: Unable to resolve opencast-services-processing-light-load/8.0.0.SNAPSHOT: missing requirement [opencast-services-processing-light-load/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-videoeditor-workflowoperation; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-videoeditor-workflowoperation/8.0.0.SNAPSHOT: missing requirement [opencast-videoeditor-workflowoperation/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.workflow.handler.workflow)(version>=8.0.0)(!(version>=9.0.0)))"]]
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
        ... 12 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-services-processing-light-load/8.0.0.SNAPSHOT: missing requirement [opencast-services-processing-light-load/8.0.0.SNAPSHOT] osgi.identity; osgi.identity=opencast-videoeditor-workflowoperation; type=osgi.bundle; version="[8.0.0.SNAPSHOT,8.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve opencast-videoeditor-workflowoperation/8.0.0.SNAPSHOT: missing requirement [opencast-videoeditor-workflowoperation/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.workflow.handler.workflow)(version>=8.0.0)(!(version>=9.0.0)))"]
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
        ... 13 more
Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve opencast-videoeditor-workflowoperation/8.0.0.SNAPSHOT: missing requirement [opencast-videoeditor-workflowoperation/8.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.opencastproject.workflow.handler.workflow)(version>=8.0.0)(!(version>=9.0.0)))"
        at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1343)
        ... 14 more
```

Edit:

This is caused by the WOH here depending on `opencast-workflow-workflowoperation`, which isn't present in the worker profile.  Adding it to the worker profile yields a dependency on `org.opencastproject.distribution.api` -> `opencast-distribution-service-api`, which yields another dependency on `org.opencastproject.presets.api`.  At this point I gave up.  You need to rework the WOH here to make it work with the worker profile.

**Comment by lwillmann** [comment]
Hi Greg, 
I've tested this locally with an admin, worker and presentation node on different port numbers and didn't run into the problem you described. 
However I've removed the dependency to the workflowoperation as it seems to be not needed ... I hope this fixes the problem you've encountered.

**Comment by lwillmann** [comment]
I don't know why the build failed ... it runs in my environment.

**Comment by gregorydlogan** [comment]
@lwillmann I've restarted the build, will probably work this time.  The previous error was
```
09:10:59 ERROR (CropServiceImpl:267) - Error extracting crop information from FFmpeg output
java.lang.IllegalArgumentException: No group with name <widthVideo>
```

**Comment by gregorydlogan** [comment]
@lwillmann I'm still seeing the exception, but I'm also seeing this with a vanilla, unpatched develop -> develop is currently broken.  I've filed #1076, and will get that done before I can continue with this.

**Pull request merged**